### PR TITLE
Fix #3326: Update Kubernetes Model to Kubernetes v1.22.1

### LINKED ISSUE #3326: Update APIs for Kubernetes 1.22
Kubernetes 1.22 (purportedly due out mid-late August 2021) is removing a number of APIs which are currently used in the Kubernetes-client.  The one I use the most is the Ingress & ingress builder, part of the extensions, which are  currently based on the extensions/v1beta1 and networking.k8s.io/v1beta1 API versions.  As of 1.22, these will no longer work.

The changes necessary & steps required are summarized in [Kubernetes API and Feature Removals](https://kubernetes.io/blog/2021/07/14/upcoming-changes-in-kubernetes-1-22/).

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description
Fix #3326 

+ Updated Kubernetes Model to Kubernetes v1.22.1

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [X] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

**Comment by manusa** [comment]
Did you check you added every new API?

I don't see anything related to:
- [Eviction policy API group](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#eviction-v1-policy)
  https://github.com/kubernetes/enhancements/issues/85

**Comment by rohanKanojia** [comment]
I'm not able to see any api resource named Eviction when I do `kubectl api-resources` on a Kubernetes 1.22 cluster.

**Comment by manusa** [comment]
Eviction is not a new resource, is the type to represent the request body for:
- `/api/v1/namespaces/{namespace}/pods/{name}/eviction`

**Comment by rohanKanojia** [comment]
Right now we do have support for Pod eviction via `evict()` method which was added in https://github.com/fabric8io/kubernetes-client/pull/2249 . However, this internally builds up a `io.fabric8.kubernetes.api.model.policy.v1beta1.Eviction`. I'm keeping it as it is and adding a new DSL method in [Evictable](https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-model-generator/kubernetes-model-policy/src/generated/java/io/fabric8/kubernetes/api/model/policy/v1beta1/Eviction.java) interface which will receive policy/v1 Eviction object:
```java
B evict(io.fabric8.kubernetes.api.model.policy.v1.Eviction eviction);
```

**Comment by rohanKanojia** [comment]
Tests seem to be failing since we don't have support for policy/v1 Eviction in Kubernetes versions < 1.22.1 . Shall I remove integration test regarding policy/v1 Eviction or shall I try adding Kubernetes 1.22 in CI suite and only run eviction test there?

**Comment by manusa** [comment]
> Tests seem to be failing since we don't have support for policy/v1 Eviction in Kubernetes versions < 1.22.1 . Shall I remove integration test regarding policy/v1 Eviction or shall I try adding Kubernetes 1.22 in CI suite and only run eviction test there?

The latter (add new CI suite and run test only in 1.22)

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/OperationSupport.java [hunk 1]
```diff
@@ -493,6 +493,23 @@ protected Status handleDeploymentRollback(String resourceUrl, DeploymentRollback
...
+   * @throws InterruptedException Interrupted Exception
+   */
+  protected void handleEviction(String evictionAsStr, URL podUrl) throws IOException, ExecutionException, InterruptedException {
+    RequestBody requestBody = RequestBody.create(JSON, evictionAsStr);
+
+    URL requestUrl = new URL(URLUtils.join(podUrl.toString(), "eviction"));
+    Request.Builder requestBuilder = new Request.Builder().post(requestBody).url(requestUrl);
+    handleResponse(requestBuilder, null, Collections.emptyMap());
+  }
+
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/OperationSupport.java:512]: - If this is specific to Pods, why did we add it to the base OperationSupport class?
- Can we accept a HasMetadata as parameter and do the conversion here?

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/core/v1/PodOperationsImpl.java [hunk 2]
```diff
@@ -280,6 +278,30 @@ public Boolean evict() {
     }
   }
 
+  @Override
+  public Boolean evict(io.fabric8.kubernetes.api.model.policy.v1.Eviction eviction) {
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/core/v1/PodOperationsImpl.java:282]: If we had a `private Boolean evict(HasMetadata eviction)` method, we could reuse most of the exception handling boiler-plate both for this method and the `evictThis` method.

File: .github/workflows/e2e-tests.yml [hunk 3]
```diff
@@ -52,14 +52,14 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        kubernetes: [v1.20.1,v1.19.2,v1.12.0]
+        kubernetes: [v1.22.1,v1.20.1,v1.19.2,v1.14.0]
```
Comment (manusa) [inline .github/workflows/e2e-tests.yml:55]: Why the bump in the low range?

Comment (rohanKanojia) [reply .github/workflows/e2e-tests.yml:55]: minikube v1.23.0 's lowest supported version is 1.14.0

**Review by manusa — APPROVED** [review]

**Pull request merged**